File: mg01/game/tech/Assets/Script/Org/Module/CCModule.cs
/* ==============================================================================
 * CCModule
 * @author jr.zeng
 * 2016/6/22 16:01:40
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


namespace mg.org
{
    public class CCModule : BaseObject, ISubject
    {

        protected bool m_isOpen = false;
        protected bool m_schUpdated = false;
        protected Subject m_notifier = null;

        public CCModule()
        {


        }

        public void Setup(params object[] params_)
        {
            if (m_isOpen) return;
            m_isOpen = true;

            __Setup(params_);
            SetupEvent();
        }

        public void Clear()
        {
            if (!m_isOpen) return;
            m_isOpen = false;

            ClearEvent();
            UnscheduleUpdate();
            DetachAll();

            __Clear();

            NotifyDeactive();
        }


        virtual protected void __Setup(params object[] params_)
        {

        }

        virtual protected void __Clear()
        {


        }


        virtual protected void SetupEvent()
        {

        }

        virtual protected void ClearEvent()
        {

        }

        //-------∽-★-∽------∽-★-∽--------∽-★-∽数据管理∽-★-∽--------∽-★-∽------∽-★-∽--------//


        //-------∽-★-∽------∽-★-∽--------∽-★-∽观察者∽-★-∽--------∽-★-∽------∽-★-∽--------//

       
        
        public void Attach(string type_, CALLBACK_1 callback_, object refer_)
        {
            if (m_notifier == null)
                m_notifier = CCApp.subject;
            m_notifier.Attach(type_, callback_, refer_);
        }

        public void Detach(string type_, CALLBACK_1 callback_)
        {
            if (m_notifier == null)
                return;
            m_notifier.Detach(type_, callback_);
        }

        public void DetachByType(string type_)
        {
            if (m_notifier == null)
                return;
            m_notifier.DetachByType(type_);
        }

        public void DetachAll()
        {
            if (m_notifier == null)
                return;
            m_notifier.DetachAll();
        }

        public bool Notify(string type_, object data_ = null)
        {
            if (m_notifier == null)
                return false;
            return m_notifier.Notify(type_, data_);
        }

        public bool NotifyEvent(SubjectEvent evt_)
        {
            if (m_notifier == null)
                return false;
            return m_notifier.NotifyEvent(evt_);
        }

        public bool NotifyWithEvent(string type_, object data_ = null)
        {
            if (m_notifier == null)
                return false;
            return m_notifier.NotifyWithEvent(type_, data_);
        }

        public void ClearNotifier()
        {
            if (m_notifier == null)
                return;
            m_notifier.DetachAll();
        }



        //-------∽-★-∽------∽-★-∽--------∽-★-∽Schedule∽-★-∽--------∽-★-∽------∽-★-∽--------//

        protected void ScheduleUpdate()
        {
            if (m_schUpdated)
                return;
            m_schUpdated = true;
            CCApp.SchUpdate(Step);
        }

        protected void UnscheduleUpdate()
        {
            if (!m_schUpdated)
                return;
            m_schUpdated = false;
            CCApp.UnschUpdate(Step);
        }


        virtual public void Step(float dt_)
        {

        }


    }
}

File: mg01/game/tech/Assets/Resources/LuaScript/src/orgu/module/CCModule.lua
--CCModule
--@author jr.zeng
--2017年9月29日 下午4:28:41
local modname = "CCModule"
--==================global reference=======================

--===================namespace========================
local ns = "org"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using)
--===================module property========================
local super = nil	--父类
_ENV[modname] = class(modname, super)
CCModule = _ENV[modname]
local CCModule = CCModule

local Refer = Refer
local App = App
--===================module content========================

function CCModule:__ctor()
    
    self.m_is_open = false
    self.m_notifier = false
end

function CCModule:setup(...)

    if self.m_is_open then
        return end
    
    self:__setup(...)
    self:setup_event()
    
    self.m_is_open = true    --有可能在__setup里调ctor导致置false,所以放后面
end


function CCModule:clear()

    if not self.m_is_open then
        return end
    self.m_is_open = false

    self:clear_event()
	safe_call(self.__clear, self)
    
    self:detachAll()
	Refer.notifyDeactive(self)
    
end

function CCModule:__setup(...)

end

function CCModule:__clear()

end

function CCModule:setup_event()

end

function CCModule:clear_event()

end


--//-------~★~-------~★~-------~★~数据管理~★~-------~★~-------~★~-------//


--//-------~★~-------~★~-------~★~数据操作~★~-------~★~-------~★~-------//

--//-------~★~-------~★~-------~★~观察者~★~-------~★~-------~★~-------//

--//-------~★~监听别人~★~-------//

--@notifer 需要监听的派发器，缺省时为公共那个
function CCModule:add_listener(type, fun, target, notifer)
	notifer = notifer or App.notifer
	notifer:attach(type, fun, target, self)
end


function CCModule:remove_listener(type, fun, target, notifer)
	notifer = notifer or App.notifer
	notifer:detach(type, fun, target)
end


function CCModule:attach_update(fun, target)
	App:attach_update(fun, target, self)
end

function CCModule:detach_update(fun, target)
	App:detach_update(fun, target)
end

--//-------~★~被别人监听~★~-------//

--添加监听
function CCModule:attach(type, fun, target, refer)
    if not self.m_notifier then
        self.m_notifier = App.notifer  end
    self.m_notifier:attach(type, fun, target, refer)
end

--移除监听
function CCModule:detach(type, fun, target)
	if not self.m_notifier then
		self.m_notifier = App.notifer  end
	self.m_notifier:detach(type, fun, target)
end

function CCModule:detachByType(type)
	if not self.m_notifier then
		self.m_notifier = App.notifer  end
	self.m_notifier:detachByType(type)
end

function CCModule:detachByType2(type, target)
	if not self.m_notifier then
		self.m_notifier = App.notifer  end
	self.m_notifier:detachByType2(type, target)
end

--派发
function CCModule:notify(name, ...)
	if not self.m_notifier then
		self.m_notifier = App.notifer  end
	self.m_notifier:notify(name, ...)
end

--用事件派发
function CCModule:notifyWithEvent(type, data)
	if not self.m_notifier then
		self.m_notifier = App.notifer  end
	self.m_notifier:notifyWithEvent(type, data)
end

function CCModule:detachAll()
    if self.m_notifier and
        self.m_notifier ~= App.notifer then --不是公共派发器
        self.m_notifier:detachAll()
    end
end

return CCModule